Controller/src/main/java/org/apidb/irods/BuildEventsFile.java

```diff
@@ -22,82 +19,106 @@
+        throw new RuntimeException(String.format(
+          "Called by wrong datastore %s. Expecting %s.",
+          datasetStoreID,
+          dsStore.getId()
+        ));
       }
 
       // Collect a subset of the event files from the events folder in the
       // datastore.
-      UserDatasetStoreAdaptor dsAdaptor = dsSession.getUserDatasetStoreAdaptor();
-      Long lastHandledEventId = null;
-      try (DatabaseInstance appDb = new DatabaseInstance(handler.getModelConfig().getAppDB(), WdkModel.DB_INSTANCE_APP, true)) {
-          DataSource appDbDataSource = appDb.getDataSource();
-          lastHandledEventId = handler.findLastHandledEvent(appDbDataSource, handler.getUserDatasetSchemaName());
-      }
-      catch (Exception e) {
-        throw new WdkModelException(e);
+      var  dsAdaptor = dsSession.getUserDatasetStoreAdaptor();
+      Long lastHandledEventId;
+      try (
+        DatabaseInstance appDb = new DatabaseInstance(
+          handler.getModelConfig().getAppDB(),
+          WdkModel.DB_INSTANCE_APP,
+          true
+        )
+      ) {
+        lastHandledEventId = handler.findLastHandledEvent(appDb.getDataSource());
       }
 
       // Read the contents of recent json formatted event files into JSON
       // objects and collect those JSON objects into a JSON array.
-      //TODO - eventJsonArray could produce big memory footprint if we handle large number of event files.
-      for(Path eventFile : dsSession.getRecentEvents(EVENTS_DIR, lastHandledEventId == null ? 0L : lastHandledEventId)) {
-        if(eventFile.getFileName().toString().endsWith(".json")) {
-          String event = dsAdaptor.readFileContents(eventFile);
-          JSONObject eventJson = new JSONObject(event);
-          eventJsonArray.put(eventJson);
+      //
+      // TODO: eventJsonArray could produce big memory footprint if we handle
+      //       large number of event files.
+      var eventJsonArray = new LinkedList<UDEvent>();
+      var eventFiles = dsSession.getRecentEvents(
+        EVENTS_DIR,
+        lastHandledEventId == null ? 0L : lastHandledEventId
+      );
+
+      for (var eventFile : eventFiles) {
+        if (eventFile.getFileName().toString().endsWith(".json")) {
+          eventJsonArray.add(EventParser.parseSingle(dsAdaptor.readFileContents(eventFile)));
         }
       }
-      logger.info("Events JSON array:" + eventJsonArray.toString(2));
+
+      LOG.info("Events JSON array:" + JsonUtil.prettyPrint(eventJsonArray));
 
       // Process the events
-      handler.handleEventList(UserDatasetEventArrayHandler.parseEventsArray(eventJsonArray));
+      handler.handleEventList(UserDatasetEventSync.parseEventsArray(eventJsonArray));
+    } catch (Exception e) {
+      throw new WdkModelException(e);
     }
   }
 
+  private static void cleanupMode(String projectID) throws WdkModelException {
+    new UserDatasetEventCleanup(projectID).cleanupFailedInstalls();
```

Comment:
i am confused by why the code for sync and for cleanup are not parallel in the way they get the list of events to process.   in one case, that logic is in-line here.   in the other it is in UserDatasetEventCleanup.   the main difference seems to be just how they find the first event to start looking at?

Reply:
The cleanup process gets the list of events from the DB and doesn't need iRODS for anything